include/core/gate.hpp

```diff
@@ -0,0 +1,277 @@
+     * @brief Mutable Iterator for iterating through control or target lines
+     *
+
+     */
+        typedef boost::transform_iterator<transform_line, boost::filter_iterator<filter_line, line_container::iterator>> iterator;
+
+        /**
+     * @brief Constant Iterator for iterating through control or target lines
+     *
+
+     */
+        typedef boost::transform_iterator<transform_line, boost::filter_iterator<filter_line, line_container::const_iterator>> const_iterator;
+
+    public:
+        /**
+     * @brief Default constructor
+     *
+     * Initializes private data
+     *
+
+     */
+        gate();
+
+        /**
+     * @brief Copy Constructor
+     *
+     * Initializes private data and copies gate
+     *
+     * @param other Gate to be assigned
+     *
+
+     */
+        gate(const gate& other);
+
+        /**
+     * @brief Default deconstructor
+     *
+     * Clears private data
+     *
+
+     */
+        virtual ~gate();
+
+        /**
+     * @brief Assignment operator
+     *
+     * @param other Gate to be assigned
+     *
+     * @return Pointer to instance
+     *
+
+     */
+        gate& operator=(const gate& other);
+
+        /**
+     * @brief Start iterator for accessing control lines. 
+     *
+     * Returns The start iterator of the line_container for accessing control lines. 
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator begin_controls() const;
+
+        /**
+     * @brief End iterator for accessing control lines (const).
+     *
+     * Returns The end iterator of the line_container for accessing control lines (const).
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator end_controls() const;
+
+        /**
+     * @brief Start iterator for accessing target lines (const).
+     *
+     * Returns The start iterator of the line_container for accessing target lines (const). 
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator begin_targets() const;
+
+        /**
+     * @brief End iterator for accessing target lines (const).
+     *
+     * Returns The end iterator of the line_container for accessing target lines (const). 
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator end_targets() const;
+
+        /**
+     * @brief Adds a control line to the gate
+     *
+     * @param c control line to add
+     *
+
+
+     */
+        virtual void add_control(line c);
+
+        /**
+     * @brief Adds a target to the desired line
+     *
+     * @param l target line 
+     *
+
+
+     */
+        virtual void add_target(line l);
+
+        /**
+     * @brief Sets the type of the target line(s)
+     *
+     * @param t target type
+     *
+
+     */
+        virtual void set_type(const std::any& t);
+
+        /**
+     * @brief Returns the type of the target line(s)
+     *
+     * @return target type
+     *
+
+     */
+        [[nodiscard]] virtual const std::any& type() const;
+
+    private:
+        struct priv;
+        priv* const d;
```

Comment:
why does this class have an extra layer of indirection?
just make the `gate` class a `struct` that has the `controls`, `targets` and `type` members.
makes the class usage so much easier and should allow the use of range-based for loops and stuff like that in the places where gates are added, manipulated and simulated.